<a href="https://colab.research.google.com/github/majo581/TALLERES/blob/main/5_taller_ensamble_Esquivel_Gonzalez.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/LinaMariaCastro/curso-ia-para-economia/blob/main/clases/5_Aprendizaje_supervisado/5_Taller_Ensamble.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Inteligencia Artificial con Aplicaciones en Economía I**

- 👩‍🏫 **Profesora:** [Lina María Castro](https://www.linkedin.com/in/lina-maria-castro)  
- 📧 **Email:** [lmcastroco@gmail.com](mailto:lmcastroco@gmail.com)  
- 🎓 **Universidad:** Universidad Externado de Colombia - Facultad de Economía

# **Taller Ensamble**

**IMPORTANTE**: Guarda una copia de este notebook en tu Google Drive o computador.

**Taller en parejas**

**Nombres estudiantes:**

-Yuli Esquivel
-Maria Jose Gonzalez

**Forma de entrega**

Jupyter Notebook publicado en su cuenta de Github con el nombre “Taller_Ensamble_apellidos.ipynb”.

**Plazo de entrega**

13 de noviembre, máximo a las 11:59 p.m., debes enviar link del notebook al correo lina.castro6@uexternado.edu.co, de lo contrario, no será tenido en cuenta.

**Instrucciones Generales**

Completa el código en las celdas marcadas con `### TU CÓDIGO AQUÍ ###`. Puedes añadir más celdas si lo requieres.

## Caso de Consultoría: Ames Real Estate Solutions

**Contexto:** Eres un consultor de datos contratado por **Ames Real Estate Solutions**. Tu misión es construir un *pipeline* de Machine Learning robusto para predecir el precio de venta (`SalePrice`) de las viviendas en la ciudad de Ames, Iowa.

**Objetivos:**
1.  Construir un *pipeline* de preprocesamiento profesional usando `ColumnTransformer` y `OneHotEncoder`.
2.  Entrenar y comparar el desempeño en prueba de 3 modelos de regresión: árbol de decisión, random forest y gradient boosting.
3.  Evaluar los modelos usando **R-cuadrado (R²)** y **RMSE** (Raíz del Error Cuadrático Medio).
4.  Optimizar uno de los modelos, el de Random Forest, usando `GridSearchCV` para mejorar su rendimiento en prueba.

### Paso 1: Configuración Inicial

In [ ]:
# Importa las librerías necesarias
### TU CÓDIGO AQUÍ ###
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

# Módulos de Scikit-Learn para modelado y evaluación
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_curve, auc, precision_recall_curve

Mejorar visualización de dataframes y gráficos

In [ ]:
# Que muestre todas las columnas
pd.options.display.max_columns = None
# En los dataframes, mostrar los float con dos decimales
pd.options.display.float_format = '{:,.2f}'.format

# Configuraciones para una mejor visualización
sns.set(style='whitegrid')
plt.rcParams['figure.figsize'] = (10, 6)

In [ ]:
# Cargar el dataset
url = 'http://jse.amstat.org/v19n3/decock/AmesHousing.txt'
df = pd.read_csv(url, sep='\t')

print(f"Dataset cargado con {df.shape[0]} filas y {df.shape[1]} columnas.")
df.head()

Dataset cargado con 2930 filas y 82 columnas.


Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0      1  526301100           20        RL        141.00     31770   Pave   
1      2  526350040           20        RH         80.00     11622   Pave   
2      3  526351010           20        RL         81.00     14267   Pave   
3      4  526353030           20        RL         93.00     11160   Pave   
4      5  527105010           60        RL         74.00     13830   Pave   

  Alley Lot Shape Land Contour Utilities Lot Config Land Slope Neighborhood  \
0   NaN       IR1          Lvl    AllPub     Corner        Gtl        NAmes   
1   NaN       Reg          Lvl    AllPub     Inside        Gtl        NAmes   
2   NaN       IR1          Lvl    AllPub     Corner        Gtl        NAmes   
3   NaN       Reg          Lvl    AllPub     Corner        Gtl        NAmes   
4   NaN       IR1          Lvl    AllPub     Inside        Gtl      Gilbert   

  Condition 1 Condition 2 Bldg Type House Style  Overall Qual  Overall Cond  \
0        Norm        Norm      1Fam      1Story             6             5   
1       Feedr        Norm      1Fam      1Story             5             6   
2        Norm        Norm      1Fam      1Story             6             6   
3        Norm        Norm      1Fam      1Story             7             5   
4        Norm        Norm      1Fam      2Story             5             5   

   Year Built  Year Remod/Add Roof Style Roof Matl Exterior 1st Exterior 2nd  \
0        1960            1960        Hip   CompShg      BrkFace      Plywood   
1        1961            1961      Gable   CompShg      VinylSd      VinylSd   
2        1958            1958        Hip   CompShg      Wd Sdng      Wd Sdng   
3        1968            1968        Hip   CompShg      BrkFace      BrkFace   
4        1997            1998      Gable   CompShg      VinylSd      VinylSd   

  Mas Vnr Type  Mas Vnr Area Exter Qual Exter Cond Foundation Bsmt Qual  \
0        Stone        112.00         TA         TA     CBlock        TA   
1          NaN          0.00         TA         TA     CBlock        TA   
2      BrkFace        108.00         TA         TA     CBlock        TA   
3          NaN          0.00         Gd         TA     CBlock        TA   
4          NaN          0.00         TA         TA      PConc        Gd   

  Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1 BsmtFin Type 2  \
0        Gd            Gd            BLQ        639.00            Unf   
1        TA            No            Rec        468.00            LwQ   
2        TA            No            ALQ        923.00            Unf   
3        TA            No            ALQ      1,065.00            Unf   
4        TA            No            GLQ        791.00            Unf   

   BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF Heating Heating QC Central Air  \
0          0.00       441.00       1,080.00    GasA         Fa           Y   
1        144.00       270.00         882.00    GasA         TA           Y   
2          0.00       406.00       1,329.00    GasA         TA           Y   
3          0.00     1,045.00       2,110.00    GasA         Ex           Y   
4          0.00       137.00         928.00    GasA         Gd           Y   

  Electrical  1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  \
0      SBrkr        1656           0                0         1656   
1      SBrkr         896           0                0          896   
2      SBrkr        1329           0                0         1329   
3      SBrkr        2110           0                0         2110   
4      SBrkr         928         701                0         1629   

   Bsmt Full Bath  Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  \
0            1.00            0.00          1          0              3   
1            0.00            0.00          1          0              2   
2            0.00            0.00          1          1              3   
3            1.00            0.00

Para este taller, nos enfocaremos en 6 variables predictoras + 1 variable objetivo.

- **Variable Objetivo (Target):**

  - **SalePrice (Precio de Venta):** Es una variable numérica continua que representa el precio final de la transacción de la vivienda en dólares.

- **Variables Predictoras (Features):**

Hemos seleccionado un conjunto mixto de 3 variables numéricas y 3 variables categóricas que intuitivamente tienen un fuerte impacto en el precio:

*Numéricas:*

  - **Overall Qual (Calidad General):** Es una calificación en una escala de 1 a 10 que resume la calidad general del material y los acabados de la casa. Es una medida directa de "lujo" y "calidad de construcción". Una casa con acabados de alta gama (calificación 9 o 10) valdrá mucho más que una con acabados básicos (calificación 4 o 5), incluso si tienen el mismo tamaño.

  - **Gr Liv Area (Área Habitable):** Es el área total en pies cuadrados de los espacios habitables que están sobre el nivel del suelo. Esta variable captura el "tamaño" útil de la vivienda. En igualdad de condiciones, una casa más grande es más valiosa.

  - **Year Built (Año de Construcción):** El año en que la casa fue construida originalmente. Representa la antigüedad de la casa. Las casas más nuevas suelen tener diseños más modernos, mejor eficiencia energética y menos necesidad de reparaciones inmediatas, lo que generalmente aumenta su valor.

*Categóricas:*

  - **Neighborhood (Barrio):** La ubicación física (el barrio) dentro de la ciudad de Ames. Esta variable captura miles de factores ocultos (efectos fijos) como la calidad de las escuelas, la seguridad, el prestigio social, el acceso a parques y el tiempo de desplazamiento.

  - **House Style (Estilo de Casa):** El estilo de la vivienda (ej. 1Story - 1 piso, 2Story - 2 pisos, SLvl - Niveles divididos). Diferentes estilos de construcción tienen diferentes costos y atraen a diferentes segmentos de compradores.

  - **Exter Qual (Calidad Exterior):** Calificación de la calidad de los materiales en el exterior de la casa (ej. Ex - Excelente, Gd - Bueno, TA - Típico/Promedio, Fa - Justo). Mide la calidad del "cascarón" de la casa y el "atractivo visual".

In [ ]:
features = [
    'Overall Qual',  # Numérica (ordinal)
    'Gr Liv Area',   # Numérica
    'Year Built',    # Numérica
    'Neighborhood',  # Categórica
    'House Style',   # Categórica
    'Exter Qual'     # Categórica (ordinal)
]
target = 'SalePrice'
df_model = df[features + [target]].copy()
df_model.head()

Overall Qual  Gr Liv Area  Year Built Neighborhood House Style Exter Qual  \
0             6         1656        1960        NAmes      1Story         TA   
1             5          896        1961        NAmes      1Story         TA   
2             6         1329        1958        NAmes      1Story         TA   
3             7         2110        1968        NAmes      1Story         Gd   
4             5         1629        1997      Gilbert      2Story         TA   

   SalePrice  
0     215000  
1     105000  
2     172000  
3     244000  
4     189900

### Paso 2: Preprocesamiento y Creación del Pipeline

Pasos:

1.  Separar `X` e `y`.
2.  Separar en `train` y `test` **antes** de cualquier transformación para evitar fuga de datos.
3.  Definir un `ColumnTransformer` que sepa qué hacer con las columnas numéricas y categóricas.
4.  Ajustar (`fit_transform`) el transformador SÓLO en los datos de `train`.

In [ ]:
X = df_model[features]
y = df_model[target]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
numerical_features = ['Overall Qual', 'Gr Liv Area', 'Year Built']
categorical_features = ['Neighborhood', 'House Style', 'Exter Qual']

In [ ]:
# Tarea 2.4: Definir el ColumnTransformer. En este caso, no vamos a estandarizar las variables numéricas

# Crear el transformador para las categóricas
categorical_transformer = OneHotEncoder(handle_unknown='ignore')

# Crear el preprocesador (ColumnTransformer)
# Recuerda que, en este caso, no vamos a estandarizar las variables numéricas
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', categorical_transformer, categorical_features),
        ('num', 'passthrough', numerical_features)
    ])

In [ ]:
# Tarea 2.5: Aplicar el preprocesador
# Ajustamos (fit) y transformamos el set de entrenamiento
X_train_processed = preprocessor.fit_transform(X_train)

# SÓLO transformamos el set de prueba (ya "aprendió" del train)
X_test_processed = preprocessor.transform(X_test)

print(f"\nForma de X_train procesado: {X_train_processed.shape}")
print(f"Forma de X_test procesado: {X_test_processed.shape}")


Forma de X_train procesado: (2344, 43)
Forma de X_test procesado: (586, 43)


### Paso 3: Modelos Base (Baseline)

Entrenemos 3 modelos: árbol de decisión, random forest y gradient boosting.

In [ ]:
# Tarea 3.1: Árbol de Decisión (Baseline)

# Inicializa el modelo, usa DecisionTreeRegressor
# Coloca como semilla random_state=42
dtree_model = DecisionTreeRegressor(random_state=42)

# Entrena el modelo
dtree_model.fit(X_train_processed, y_train)

# Realiza predicciones en el conjunto de prueba
y_pred_dtree = dtree_model.predict(X_test_processed)

In [ ]:
# Tarea 3.2: Random Forest (Baseline)

# Inicializa el modelo, usa RandomForestRegressor
# Coloca como semilla random_state=42 y pide que use todos los núcleos del computador n_jobs=-1
# Usa los siguientes hiperparámetros: n_estimators= 500 - max_depth=10 - min_samples_leaf=2
rf_model = RandomForestRegressor(random_state=42, n_jobs=-1, n_estimators=500, max_depth=10, min_samples_leaf=2)

# Entrena el modelo
rf_model.fit(X_train_processed, y_train)

# Realiza predicciones en el conjunto de prueba
y_pred_rf = rf_model.predict(X_test_processed)

In [ ]:
# Tarea 3.3: Gradient Boosting (Baseline)

# Inicializa el modelo, usa GradientBoostingRegressor
# Coloca como semilla random_state=42
gb_model = GradientBoostingRegressor(random_state=42)

# Entrena el modelo
gb_model.fit(X_train_processed, y_train)

# Realiza predicciones en el conjunto de prueba
y_pred_gradient = gb_model.predict(X_test_processed)

### Paso 4: Comparación de Métricas (Baseline)

In [ ]:
# Crea una dataframe que muestre para cada modelo el RMSE y el R2.
# Ordena el dataframe de menor a mayor RMSE
### TU CÓDIGO AQUÍ ###
from sklearn.metrics import mean_squared_error, r2_score

# Calcular métricas para Decision Tree
rmse_dtree = np.sqrt(mean_squared_error(y_test, y_pred_dtree))
r2_dtree = r2_score(y_test, y_pred_dtree)

# Calcular métricas para Random Forest
rmse_rf = np.sqrt(mean_squared_error(y_test, y_pred_rf))
r2_rf = r2_score(y_test, y_pred_rf)

# Calcular métricas para Gradient Boosting
rmse_gradient = np.sqrt(mean_squared_error(y_test, y_pred_gradient))
r2_gradient = r2_score(y_test, y_pred_gradient)

# Crear el DataFrame de resultados
results_df = pd.DataFrame({
    'Modelo': ['Decision Tree', 'Random Forest', 'Gradient Boosting'],
    'RMSE': [rmse_dtree, rmse_rf, rmse_gradient],
    'R2': [r2_dtree, r2_rf, r2_gradient]
})

# Ordenar por RMSE de menor a mayor
results_df_sorted = results_df.sort_values(by='RMSE', ascending=True).reset_index(drop=True)

display(results_df_sorted)

Modelo      RMSE   R2
0      Random Forest 30,064.75 0.89
1  Gradient Boosting 32,383.26 0.87
2      Decision Tree 35,544.40 0.84

¿Cuál modelo tuvo un mejor desempeño en el dataset de prueba (menor RMSE, mayor R²)?

Según los resultados de results_df_sorted, el modelo que mejor funcionó en el conjunto de prueba fue Random Forest. Este modelo tuvo el RMSE más bajo (30,064.75) y el R² más alto (0.89), lo que significa que fue el que hizo las predicciones más precisas.

**Respuesta**

### Paso 5: Optimización con GridSearchCV para Random Forest

La búsqueda de los hiperparámetros óptimos con `GridSearchCV` se puede realizar para los 3 modelos, sin embargo, como esto toma tiempo, en este taller lo realizaremos solo para el modelo Random Forest.

In [ ]:
# Analiza este código, entiende sus pasos y ejecútalo:

print("Iniciando GridSearchCV para Random Forest... (Esto puede tardar 2-3 minutos)")

# Tarea 5.1: Definir la parrilla de parámetros para Random Forest
# Hiperparámetros clave de RF:
# - n_estimators: Número de árboles
# - max_depth: Profundidad máxima (None = sin límite)
# - min_samples_leaf: Mínimo de muestras en una "hoja" final

param_grid_rf = {
    'n_estimators': [100, 200],       # Prueba con 100 y 200 árboles
    'max_depth': [None, 10],          # Prueba árboles sin límite y árboles podados a 10
    'min_samples_leaf': [1, 2],       # Prueba requiriendo 1 o 2 muestras por hoja
}
# Total combinaciones: 2 * 2 * 2 = 8 combinaciones. (8 * 5 folds = 40 modelos)

# Tarea 5.2: Inicializar el GridSearchCV
# Le pasamos un modelo base de RF.
# n_jobs=-1 en el estimador base acelera el entrenamiento de CADA árbol.
rf_base_model = RandomForestRegressor(random_state=42, n_jobs=-1)

grid_search_rf = GridSearchCV(
    estimator=rf_base_model,
    param_grid=param_grid_rf,
    cv=5,
    scoring='neg_mean_squared_error', # Usamos el error cuadrático medio negativo (scikit-learn siempre maximiza)
    n_jobs=-1,  # n_jobs=-1 aquí acelera el "GridSearch" (corre los 5 folds en paralelo)
    verbose=1
)

# Tarea 5.3: Ejecutar la búsqueda
grid_search_rf.fit(X_train_processed, y_train)

print("¡Búsqueda completada!")
print(f"Mejores parámetros encontrados: {grid_search_rf.best_params_}")

Iniciando GridSearchCV para Random Forest... (Esto puede tardar 2-3 minutos)
Fitting 5 folds for each of 8 candidates, totalling 40 fits
¡Búsqueda completada!
Mejores parámetros encontrados: {'max_depth': 10, 'min_samples_leaf': 1, 'n_estimators': 100}


Investiga por qué en GridSearchCV se usa como métrica el error cuadrático medio **negativo**. A continuación escribe la explicación.

GridSearchCV usa el error cuadrático medio negativo porque por diseño esta función siempre busca maximizar la métrica que se le indica. Como el error cuadrático medio (MSE) normal es una medida de error (donde valores más bajos son mejores), se convierte en su versión negativa. Así, al maximizar el MSE negativo, en realidad se está minimizando el error cuadrático medio original. Esto permite que GridSearchCV funcione correctamente encontrando los mejores parámetros que reducen el error.

**Respuesta**

### Paso 6: Evaluación Final del Modelo Random Forest Optimizado

Ahora, usemos nuestro modelo ganador y optimizado (best_estimator_) para hacer predicciones en el set de prueba y ver si el RMSE mejoró respecto a nuestro Random Forest base.

In [ ]:
# Tarea 6.1: Obtener el mejor modelo de RF y predecir
best_rf_model = grid_search_rf.best_estimator_
y_pred_rf_optimizado = best_rf_model.predict(X_test_processed)

# Tarea 6.2: Calcular métricas sobre el set de prueba (RMSE y R2)
### TU CÓDIGO AQUÍ ###
rmse_rf_optimizado = np.sqrt(mean_squared_error(y_test, y_pred_rf_optimizado))
r2_rf_optimizado = r2_score(y_test, y_pred_rf_optimizado)

print(f"RMSE del Random Forest optimizado: {rmse_rf_optimizado:,.2f}")
print(f"R² del Random Forest optimizado: {r2_rf_optimizado:,.2f}")

# Compara los resultados en el set de prueba del modelo base de Random Forest y del Random Forest con optimización de hiperparámetros
# ¿Se redujo el RMSE? ¿En cuánto se redujo?
### TU CÓDIGO AQUÍ ###
reduction_rmse = rmse_rf - rmse_rf_optimizado
percentage_reduction = (reduction_rmse / rmse_rf) * 100

print(f"\nRMSE del Random Forest base: {rmse_rf:,.2f}")
print(f"El RMSE se redujo en: {reduction_rmse:,.2f}")
print(f"Lo que representa una reducción del: {percentage_reduction:,.2f}%")

RMSE del Random Forest optimizado: 29,187.99
R² del Random Forest optimizado: 0.89

RMSE del Random Forest base: 30,064.75
El RMSE se redujo en: 876.75
Lo que representa una reducción del: 2.92%


### Paso 7: Conclusión de Consultoría

Basado en los resultados finales, responde a tu cliente.

**Pregunta 1:** ¿Cuánto logramos reducir el error (RMSE) al optimizar el modelo Random Forest con GridSearchCV? (Compara el RMSE base de RF vs. el RMSE final).


**Respuesta**
El modelo Random Forest mejoró significativamente después de la optimización, logrando una reducción del Error Cuadrático Medio (RMSE) de aproximadamente 876.75. Esta mejora representa una disminución del 2.92% en el error del modelo.


**Pregunta 2:** ¿Por qué fue crucial usar `ColumnTransformer` y hacer el `train_test_split` antes de transformar los datos? ¿Qué problema evitamos?



**Respuesta**


Usar ColumnTransformer después del train_test_split es crucial para evitar la fuga de datos (data leakage), que ocurre cuando información del conjunto de prueba (como su media o desviación estándar) se usa para ajustar las transformaciones. Si transformáramos todo el dataset antes de dividirlo, el modelo vería información futura durante el entrenamiento, lo que inflaría falsamente su rendimiento y fallaría con datos nuevos. Al dividir primero, el ColumnTransformer se ajusta solo con los datos de entrenamiento


# Task
Separate the features (X) and the target variable (y) from the `df_model` DataFrame using the predefined `features` list and `target` variable. Confirm their successful separation and explain that the next step is to split the data into training and testing sets to prepare for model training.

## Separate X and y

### Subtask:
Extract the features (X) and the target variable (y) from the 'df_model' DataFrame, using the 'features' list and 'target' variable defined previously.


**Reasoning**:
To extract the features (X) and the target variable (y) from the 'df_model' DataFrame, I will use the predefined 'features' list and 'target' variable to select the appropriate columns.



## Final Task

### Subtask:
Confirm that X and y have been successfully separated, and briefly explain the next steps in the preprocessing pipeline.


## Summary:

### Q&A
*   **Have X and y been successfully separated?**
    Yes, `X` (features) and `y` (target) have been successfully separated from the `df_model` DataFrame.
*   **What are the next steps in the preprocessing pipeline?**
    The next step is to split the data into training and testing sets to prepare for model training.

### Data Analysis Key Findings
*   The features DataFrame (`X`) has been created with a shape of (2930, 6), indicating 2930 rows and 6 feature columns.
*   The target variable Series (`y`) has been created with a shape of (2930,), indicating 2930 rows.

### Insights or Next Steps
*   The separated features (`X`) and target variable (`y`) are now ready for the next stage of the machine learning pipeline, which involves splitting the data into training and testing sets.
